### ISSUE TITLE
chapter 6: Waiting on transform from /base_footprint to /map to become available before running costmap, tf error: 

### ISSUE THREAD

**Issue by stevealbertwong** [rosbook/effective_robotics_programming_with_ros#15]
I am running  roslaunch chapter6_tutorials move_base.launch on kinetic
it seems to be a issue happened before but I could not find answer on how it is solved.

https://answers.ros.org/question/224242/tf-transformation-with-move_base/
https://answers.ros.org/question/195175/waiting-on-transform-from-base_footprint-to-map-to-become-available-before-running-costmap-tf-error/
https://github.com/AaronMR/Learning_ROS_for_Robotics_Programming_2nd_edition/issues/17 

this is my log
2018-04-11 20:02:02 ⌚  steve-ThinkPad-X220 in ~/dev/catkin_ws
± |master ✓| → roslaunch chapter6_tutorials move_base.launch 
... logging to /home/steve/.ros/log/9a3117c6-3d7f-11e8-a81a-a088b4e53910/roslaunch-steve-ThinkPad-X220-9931.log
Checking log directory for disk usage. This may take awhile.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://steve-ThinkPad-X220:42859/

SUMMARY
========

PARAMETERS
 * /amcl/gui_publish_rate: 10.0
 * /amcl/kld_err: 0.05
 * /amcl/kld_z: 0.99
 * /amcl/laser_lambda_short: 0.1
 * /amcl/laser_likelihood_max_dist: 2.0
 * /amcl/laser_max_beams: 30
 * /amcl/laser_model_type: likelihood_field
 * /amcl/laser_sigma_hit: 0.2
 * /amcl/laser_z_hit: 0.5
 * /amcl/laser_z_max: 0.05
 * /amcl/laser_z_rand: 0.5
 * /amcl/laser_z_short: 0.05
 * /amcl/max_particles: 5000
 * /amcl/min_particles: 500
 * /amcl/odom_alpha1: 0.2
 * /amcl/odom_alpha2: 0.2
 * /amcl/odom_alpha3: 0.8
 * /amcl/odom_alpha4: 0.2
 * /amcl/odom_alpha5: 0.1
 * /amcl/odom_frame_id: odom
 * /amcl/odom_model_type: diff
 * /amcl/recovery_alpha_fast: 0.0
 * /amcl/recovery_alpha_slow: 0.0
 * /amcl/resample_interval: 1
 * /amcl/transform_tolerance: 0.1
 * /amcl/update_min_a: 0.5
 * /amcl/update_min_d: 0.2
 * /move_base/TrajectoryPlannerROS/acc_lim_th: 3.2
 * /move_base/TrajectoryPlannerROS/acc_lim_x: 2.5
 * /move_base/TrajectoryPlannerROS/acc_lim_y: 2.5
 * /move_base/TrajectoryPlannerROS/holonomic_robot: False
 * /move_base/TrajectoryPlannerROS/max_rotational_vel: 0.15
 * /move_base/TrajectoryPlannerROS/max_vel_theta: 0.15
 * /move_base/TrajectoryPlannerROS/max_vel_x: 0.2
 * /move_base/TrajectoryPlannerROS/min_in_place_rotational_vel: 0.01
 * /move_base/TrajectoryPlannerROS/min_in_place_vel_theta: 0.01
 * /move_base/TrajectoryPlannerROS/min_vel_theta: -0.15
 * /move_base/TrajectoryPlannerROS/min_vel_x: 0.05
 * /move_base/controller_frequency: 10.0
 * /move_base/controller_patiente: 15.0
 * /move_base/global_costmap/cost_scaling_factor: 10.0
 * /move_base/global_costmap/footprint: [[-0.2, -0.2], [-...
 * /move_base/global_costmap/global_frame: /map
 * /move_base/global_costmap/inflation_radius: 0.5
 * /move_base/global_costmap/observation_sources: scan
 * /move_base/global_costmap/obstacle_range: 2.5
 * /move_base/global_costmap/raytrace_range: 3.0
 * /move_base/global_costmap/robot_base_frame: /base_footprint
 * /move_base/global_costmap/scan/clearing: True
 * /move_base/global_costmap/scan/data_type: LaserScan
 * /move_base/global_costmap/scan/marking: True
 * /move_base/global_costmap/scan/max_obstacle_height: 0.4
 * /move_base/global_costmap/scan/min_obstacle_height: 0.0
 * /move_base/global_costmap/scan/observation_persistence: 0.0
 * /move_base/global_costmap/scan/sensor_frame: base_link
 * /move_base/global_costmap/scan/topic: /scan
 * /move_base/global_costmap/static_map: True
 * /move_base/global_costmap/update_frequency: 1.0
 * /move_base/local_costmap/cost_scaling_factor: 10.0
 * /move_base/local_costmap/footprint: [[-0.2, -0.2], [-...
 * /move_base/local_costmap/global_frame: /map
 * /move_base/local_costmap/height: 5.0
 * /move_base/local_costmap/inflation_radius: 0.5
 * /move_base/local_costmap/observation_sources: scan
 * /move_base/local_costmap/obstacle_range: 2.5
 * /move_base/local_costmap/planner_frequency: 1.0
 * /move_base/local_costmap/planner_patiente: 5.0
 * /move_base/local_costmap/publish_frequency: 2.0
 * /move_base/local_costmap/raytrace_range: 3.0
 * /move_base/local_costmap/resolution: 0.02
 * /move_base/local_costmap/robot_base_frame: /base_footprint
 * /move_base/local_costmap/rolling_window: True
 * /move_base/local_costmap/scan/clearing: True
 * /move_base/local_costmap/scan/data_type: LaserScan
 * /move_base/local_costmap/scan/marking: True
 * /move_base/local_costmap/scan/max_obstacle_height: 0.4
 * /move_base/local_costmap/scan/min_obstacle_height: 0.0
 * /move_base/local_costmap/scan/observation_persistence: 0.0
 * /move_base/local_costmap/scan/sensor_frame: base_link
 * /move_base/local_costmap/scan/topic: /scan
 * /move_base/local_costmap/static_map: False
 * /move_base/local_costmap/tranform_tolerance: 0.5
 * /move_base/local_costmap/update_frequency: 5.0
 * /move_base/local_costmap/width: 5.0
 * /rosdistro: kinetic
 * /rosversion: 1.12.13

NODES
  /
    amcl (amcl/amcl)
    map_server (map_server/map_server)
    move_base (move_base/move_base)

ROS_MASTER_URI=http://localhost:11311

process[map_server-1]: started with pid [9948]
process[amcl-2]: started with pid [9949]
process[move_base-3]: started with pid [9951]
[ INFO] [1523448893.588287083]: Loading map from image "/home/steve/dev/catkin_ws/src/chapter6_tutorials/maps/map.pgm"
[ WARN] [1523448893.680908610]: ignoring NAN in initial pose X position
[ WARN] [1523448893.683735113]: ignoring NAN in initial pose Y position
[ INFO] [1523448893.732907166]: Requesting the map...
[ WARN] [1523448893.733614535]: Request for map failed; trying again...
[ INFO] [1523448893.792339989]: Read a 4000 X 4000 map @ 0.050 m/cell
[ INFO] [1523448894.253528562]: Sending map
[ INFO] [1523448894.286647263]: Received a 4000 X 4000 map @ 0.050 m/pix

[ WARN] [1523448894.837868726]: ignoring NAN in initial pose X position
[ WARN] [1523448894.839484492]: ignoring NAN in initial pose Y position
[ INFO] [1523448894.852266034]: Initializing likelihood field model; this can take some time on large maps...
[ INFO] [1523448895.350413585]: Done initializing likelihood field model.
[ WARN] [1523448898.794508792]: Timed out waiting for transform from base_footprint to map to become available before running costmap, tf error: canTransform: target_frame map does not exist. canTransform: source_frame base_footprint does not exist.. canTransform returned after 0.101473 timeout was 0.1.
[ WARN] [1523448903.862222936]: Timed out waiting for transform from base_footprint to map to become available before running costmap, tf error: canTransform: target_frame map does not exist. canTransform: source_frame base_footprint does not exist.. canTransform returned after 0.101284 timeout was 0.1.
[ WARN] [1523448908.920670194]: Timed out waiting for transform from base_footprint to map to become available before running costmap, tf error: canTransform: target_frame map does not exist. canTransform: source_frame base_footprint does not exist.. canTransform returned after 0.101244 timeout was 0.1.
[ WARN] [1523448910.476452149]: No laser scan received (and thus no pose updates have been published) for 1523448910.476290 seconds.  Verify that data is being published on the /scan topic.
[ WARN] [1523448913.988248548]: Timed out waiting for transform from base_footprint to map to become available before running costmap, tf error: canTransform: target_frame map does not exist. canTransform: source_frame base_footprint does not exist.. canTransform returned after 0.101114 timeout was 0.1.

node graph:
![screenshot from 2018-04-13 12-31-36](https://user-images.githubusercontent.com/8135740/38720913-9c9d8fc4-3f2a-11e8-9e66-555767b5b926.png)

![37072361-f42a2b72-21fb-11e8-92ad-836125f6ce8e](https://user-images.githubusercontent.com/8135740/38721050-435a6e2c-3f2b-11e8-8316-aa533430ecec.png)

I implemented a quick fix which is run the odometry.cpp. The robot does move but in a artificially predefined way by odometry.cpp. 

![screenshot from 2018-04-12 17-54-23](https://user-images.githubusercontent.com/8135740/38720970-e20924ce-3f2a-11e8-9808-bd08f798c2ce.png)

![screenshot from 2018-04-13 12-12-19](https://user-images.githubusercontent.com/8135740/38721037-38a2ed4c-3f2b-11e8-897c-6ac11ef0d05c.png)

I wonder how could implement sendgoals.cpp and make the robot runs with a goal on map according to global and local planning with obstacles avoidance.

**Comment by stevealbertwong** [comment]
[ WARN] [1523529194.465359284, 616.269000000]: Timed out waiting for transform from base_footprint to map to become available before running costmap, tf error: Could not find a connection between 'map' and 'base_footprint' because they are not part of the same tree.Tf has two or more unconnected trees.. canTransform returned after 0.101 timeout was 0.1.
[ WARN] [1523529201.717386068, 620.854000000]: No laser scan received (and thus no pose updates have been published) for 620.854000 seconds.  Verify that data is being published on the /scan topic.
[ WARN] [1523529201.788259219, 620.897000000]: MessageFilter [target=odom ]: Dropped 100.00% of messages so far. Please turn the [ros.amcl.message_notifier] rosconsole logger to DEBUG for more information.

**Comment by stevealbertwong** [comment]
I have fixed the warnning of [ WARN] [1523448910.476452149]: No laser scan received (and thus no pose updates have been published) for 1523448910.476290 seconds. Verify that data is being published on the /scan topic.
[ WARN] [1523448913.988248548]: Timed out waiting for transform from base_footprint to map to become available before running costmap, tf error: canTransform: target_frame map does not exist. canTransform: source_frame base_footprint does not exist.. canTransform returned after 0.101114 timeout was 0.1.
by adding to launch file 
static_transform_publisher and mapped map to odom, odom to base_footprint
but i still have the issue when i give it a direction/goal, it does not move. 
i wonder its because of my config if local planner or other files ?

**Comment by Xunzhaocunzi** [comment]
I have the exactly same issue. Help please. 

In addition, when I run "roslaunch chapter6_tutorials chapter6_configuration_gazebo.launch", I got something like this, rather than the contents shown in the book: 
![screenshot from 2018-10-31 00-51-25](https://user-images.githubusercontent.com/32579634/47768695-23055000-dca7-11e8-95fb-866ef2bcbbe2.png)

For instance, my model did not have "Robot Footprint", "Global Panner", etc. 

I guess that was caused by using different "chapter6_configuration_gazebo.launch" file. 

Can authors update the files for chapter6? Thank  you.

**Comment by wx0205** [comment]
I have the exactly same issue. Help please.Could you tell me how to solve it？Thank you!
@stevealbertwong
